docs/design.md

```diff
@@ -0,0 +1,196 @@
+# Job Worker Service
+
+This document describes the design of a prototype job worker service to run arbitrary Linux processes.
+
+The diagram below shows the interactions between the components of the system. 
+These are described in the following sections.
+
+![high level diagram](high_level_diagram.png)
+
+## CLI
+
+The CLI is the main way users will interact with the worker. Users can invoke the command line application `job` to start a process, stop a process, get the process status and stream the output of a process in real time. The CLI will make calls to the GRPC API to fulfill all user commands.
+
+Below are some examples of what interacting with the CLI will be like
+
+### Starting a process
+
+Users can start processes in the worker host and are able to provide limits for memory, IO and CPU. These parameters are described more in depth in the [Resource Control](#resource-control) section. The command returns
+the UUID of the process which can be used to invoke further commands.
+
+```
+> job start –memMaxBytes=1000000 –maxReadBps=1000000 –maxWriteBps=1000000 –cpuBandwidth="200000 1000000" "scripts/release-docker.sh"
+job 9de95676-84ed-4e37-b44c-4f4d1df7c74f started
+```
+
+### Stopping a process
+
+Users can stop a process running in the worker host using the stop command and providing the job id.
+
+```
+> job stop 9de95676-84ed-4e37-b44c-4f4d1df7c74f
+job 9de95676-84ed-4e37-b44c-4f4d1df7c74f stopped
+```
+
+### Getting the status of a process
+
+Users can invoke the status command to get information about the process.
+
+```
+> job status 9de95676-84ed-4e37-b44c-4f4d1df7c74f
+{
+	"job_id": "9de95676-84ed-4e37-b44c-4f4d1df7c74f",
+	"command": "ls -la",
+	"pid": 51032,
+	"state": "Z",
+	"exit_code": 0,
+	"started_by": "hashi",
+	"started at": "2022-04-15T09:47:02Z",
+	"finished at": "2022-04-15T10:47:02Z"
+}
+```
+
+### Streaming the output of a process
+
+Users can invoke the output command to stream the output of a process in real time (a combination
+of stdout and stderr). It gets the output right from the start. The command terminates once there
+is no more output to read i.e the process is done or has been terminated.
+
+```
+> job output 9de95676-84ed-4e37-b44c-4f4d1df7c74f
+test command output line 1
+test command output line 2
+test command output line 3
+test command output line 3
+…
+```
+
+## Resource Control
+
+We will use [cgroupsv2](https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html) to assign system resources to processes.
+
+It is a requirement that the host runs Linux kernel 4.15 (or above) as that is the version when the cpu controller was
+introduced.
+
+On startup, the worker will create a worker root cgroup which will optionally contain some 
+default values that can be provided via flags. For every process, we will create a new children of the
+worker cgroup with the provided resource control values. 
+
+We will enable users to control memory, IO and CPU. We will achieve this by exposing
```

Comment:
What happens to the process if these limits are exceeded?

Reply:
That is a very good question. It will depend on what limits are exceeded.

In the case of IO and CPU limits it will take forever and as a user you will notice that the process state is "sleeping", in which you might want to stop the process and retry with more resources.

In the case of memory limits, the behaviour is:

>  If a cgroup’s memory usage reaches this limit and can’t be reduced, the OOM killer is invoked in the cgroup. Under certain circumstances, the usage may go over the limit temporarily.

I tested this and if I try to run a process without enough memory allowance the following happens:

```
dl@pop-os:# sleep 120
Killed
```
Kernel Logs:
```
Apr 21 14:29:06 pop-os kernel: [178228.853564] sleep invoked oom-killer: gfp_mask=0xcc0(GFP_KERNEL), order=0, oom_score_adj=0
Apr 21 14:29:06 pop-os kernel: [178228.853656]  oom_kill_process.cold+0xb/0x10
Apr 21 14:29:06 pop-os kernel: [178228.853968] oom-kill:constraint=CONSTRAINT_MEMCG,nodemask=(null),cpuset=/,mems_allowed=0,oom_memcg=/testing_stuff,task_memcg=/testing_stuff,task=sleep,pid=420326,uid=0
```

This behavior seems reasonable to me as long as it is documented. What are your thoughts?

Reply:
Yeah, totally. As long as the consequences have been examined and documented I'm happy.